File: src/test/kotlin/io/energyconsumptionoptimizer/mapservice/application/FloorPlanServiceImplTest.kt
package io.energyconsumptionoptimizer.mapservice.application

import io.energyconsumptionoptimizer.mapservice.domain.FloorPlan
import io.energyconsumptionoptimizer.mapservice.domain.errors.FlorPlanFormatNotValidException
import io.energyconsumptionoptimizer.mapservice.domain.ports.HouseMapRepository
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.clearAllMocks
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk

class FloorPlanServiceImplTest :
    ShouldSpec(
        {
            lateinit var repository: HouseMapRepository
            lateinit var floorPlanService: FloorPlanServiceImpl

            val svgContent = "<svg><rect width='100' height='100'/></svg>"

            beforeEach {
                repository = mockk()
                floorPlanService = FloorPlanServiceImpl(repository)
            }

            afterEach {
                clearAllMocks()
            }

            context("createFloorPlan") {
                should("create and save a floor plan with valid SVG content") {
                    val savedFloorPlan = FloorPlan(svgContent)

                    coEvery { repository.saveFloorPlan(any()) } returns savedFloorPlan

                    val result = floorPlanService.createFloorPlan(svgContent)

                    result.svgContent shouldBe svgContent
                    coVerify(exactly = 1) { repository.saveFloorPlan(match { it.svgContent == svgContent }) }
                }
                should("throw exception when SVG content is empty") {
                    val emptySvgContent = ""

                    shouldThrow<FlorPlanFormatNotValidException> {
                        floorPlanService.createFloorPlan(emptySvgContent)
                    }

                    coVerify(exactly = 0) { repository.saveFloorPlan(any()) }
                }
            }

            context("getFloorPlan") {
                should("return floor plan when it exists") {
                    val existingFloorPlan = FloorPlan(svgContent)

                    coEvery { repository.findFloorPlan() } returns existingFloorPlan

                    val result = floorPlanService.getFloorPlan()

                    result shouldNotBe null
                    result?.svgContent shouldBe svgContent
                    coVerify(exactly = 1) { repository.findFloorPlan() }
                }
                should("return null when floor plan does not exist") {
                    coEvery { repository.findFloorPlan() } returns null

                    val result = floorPlanService.getFloorPlan()

                    result shouldBe null
                    coVerify(exactly = 1) { repository.findFloorPlan() }
                }
            }
        },
    )

File: src/test/unit/application/FloorPlanServiceImpl.test.ts
import type { HouseMapRepository } from "@domain/ports/HouseMapRepository";
import type { BusinessMetrics } from "@application/outbound/BusinessMetrics";
import { FloorPlanServiceImpl } from "@application/FloorPlanServiceImpl";
import { FloorPlan } from "@domain/values/FloorPlan";
import { InvalidFloorPlanError } from "@domain/errors";
import { beforeEach, describe, expect, it } from "vitest";
import { mock, type MockProxy } from "vitest-mock-extended";
import { validFloorPlan } from "@test/domainFactories";

describe("FloorPlanServiceImpl", () => {
  let repository: MockProxy<HouseMapRepository>;
  let metrics: MockProxy<BusinessMetrics>;
  let service: FloorPlanServiceImpl;

  beforeEach(() => {
    repository = mock<HouseMapRepository>();
    metrics = mock<BusinessMetrics>();
    service = new FloorPlanServiceImpl(repository, metrics);
  });

  describe("createFloorPlan()", () => {
    it("should validate, save, record metrics, and return the floor plan for a valid SVG string", async () => {
      const svg = "<svg><rect/></svg>";

      const result = await service.createFloorPlan(svg);

      expect(result).toBeInstanceOf(FloorPlan);
      expect(repository.saveFloorPlan).toHaveBeenCalledWith(
        expect.any(FloorPlan),
      );
      expect(metrics.recordFloorPlanCreation).toHaveBeenCalledOnce();
    });

    it("should accept SVG with attributes on the opening tag", async () => {
      const svg =
        '<svg xmlns="http://www.w3.org/2000/svg"><circle r="5"/></svg>';

      const result = await service.createFloorPlan(svg);

      expect(result).toBeInstanceOf(FloorPlan);
      expect(metrics.recordFloorPlanCreation).toHaveBeenCalledOnce();
    });

    it("should accept SVG with leading whitespace", async () => {
      const result = await service.createFloorPlan("  \n<svg></svg>");

      expect(result).toBeInstanceOf(FloorPlan);
      expect(repository.saveFloorPlan).toHaveBeenCalled();
    });

    it("should return InvalidFloorPlanError for non-SVG markup without saving or recording metrics", async () => {
      const result = await service.createFloorPlan("<html><body/></html>");

      expect(result).toBeInstanceOf(InvalidFloorPlanError);
      expect(repository.saveFloorPlan).not.toHaveBeenCalled();
      expect(metrics.recordFloorPlanCreation).not.toHaveBeenCalled();
    });

    it("should return InvalidFloorPlanError for empty string without saving or recording metrics", async () => {
      const result = await service.createFloorPlan("");

      expect(result).toBeInstanceOf(InvalidFloorPlanError);
      expect(repository.saveFloorPlan).not.toHaveBeenCalled();
      expect(metrics.recordFloorPlanCreation).not.toHaveBeenCalled();
    });

    it("should return InvalidFloorPlanError for plain text without saving or recording metrics", async () => {
      const result = await service.createFloorPlan("just some text");

      expect(result).toBeInstanceOf(InvalidFloorPlanError);
      expect(repository.saveFloorPlan).not.toHaveBeenCalled();
      expect(metrics.recordFloorPlanCreation).not.toHaveBeenCalled();
    });
  });

  describe("getFloorPlan()", () => {
    it("should return the floor plan when one exists", async () => {
      const floorPlan = validFloorPlan();
      repository.findFloorPlan.mockResolvedValue(floorPlan);

      const result = await service.getFloorPlan();

      expect(result).toBe(floorPlan);
    });

    it("should return null when no floor plan has been saved yet", async () => {
      repository.findFloorPlan.mockResolvedValue(null);

      const result = await service.getFloorPlan();

      expect(result).toBeNull();
    });
  });
});
